Fixes and enhancements to Cosmos KV host implementation

kate-goldenring opened the pull request:
This contains some enhancements and a couple important bug fixes

Bug fixes:
- deletes (on instances without `store_id` configured) were not working due to using the incorrect store ID
- increment was failing due to the value never being initially set. Now this sets the value if DNE -- this is the same as Spin's other KV hosts and it is impossible to set a number counter through the `set` api which requires a string value
- gets were failing on pairs that were set with increment (because the value type was a number instead of string). Deserialized into an agnostic `Keys` structure to help with this.

Enhancements:
- ~Support for writing and querying data in non-primary regions by enabling `allow_tentative_writes`. This is needed for databases configured with global replicas if not using the primary replica~ Removing this until we enable configuring region in runtime config.

itowlson commented on crates/key-value-azure/src/store.rs:
```diff
@@ -353,15 +412,18 @@ impl Cas for CompareAndSwap {
 
 impl AzureCosmosStore {
-    async fn get_pair(&self, key: &str) -> Result<Option<Pair>, Error> {
+    async fn get_pair<F>(&self, key: &str) -> Result<Option<F>, Error>
```
Should we change the name of this function?  It seems like it's no longer limited to pairs.

rylev replied:
Maybe `get_entity`?

kate-goldenring replied:
updated to `get_entity`

itowlson commented on crates/key-value-azure/src/store.rs:
```diff
@@ -158,24 +159,35 @@ impl Store for AzureCosmosStore {
-            document_client.delete_document().await.map_err(log_error)?;
+        let document_client = self
+            .client
+            .document_client(key, &self.store_id.clone().unwrap_or(key.to_string()))
```
Sanity check... `store_id` seems to be a partition key (or PK column???) but `key` seems to be a row key - is it correct to mix them?  The docs imply the default store id is `/id` rather than the current row key, but I may be misunderstanding how the store ID gets applied.

_EDIT: oh is this aligning with `Pair`'s implementation of `CosmosEntity::partition_key`?  that feels like we should make the relationship clearer then.  Maybe improve docs on `self.store_id` too?_

kate-goldenring replied:
I updated the `store_id` inline doc to explain that `/id` is the row key:
```
    /// An optional store id to use as a partition key for all operations.
    ///
    /// If the store id not set, the store will use `/id` (the row key) as the
    /// partition key. For example, if `store.set("my_key", "my_value")` is
    /// called, the partition key will be `my_key` if the store id is set to
    /// `None`. If the store id is set to `Some("myappid/default"), the
    /// partition key will be `myappid/default`.
    store_id: Option<String>,
}
```

itowlson commented on crates/key-value-azure/src/store.rs:
```diff
@@ -216,24 +228,67 @@ impl Store for AzureCosmosStore {
+        match self
             .client
-            .document_client(key.clone(), &self.store_id)
+            .document_client(&key, &self.store_id.clone().unwrap_or(key.to_string()))
```
same question about whether `key` is the right fallback here _EDIT: and same edit!_

itowlson commented on crates/key-value-azure/src/store.rs:
```diff
@@ -216,24 +228,67 @@ impl Store for AzureCosmosStore {
+                        .allow_tentative_writes(TentativeWritesAllowance::Allow)
+                        .await
+                    {
+                        if e.as_http_error()
```
I found the nesting of similar-looking code hard to follow here. It would be great to pull out some chunks and/or helpers here but I appreciate it may not be easy given retries etc!

itowlson commented on crates/key-value-azure/src/store.rs:
```diff
@@ -274,8 +329,9 @@ impl CompareAndSwap {
 impl Cas for CompareAndSwap {
-    /// `current` will fetch the current value for the key and store the etag for the record. The
-    /// etag will be used to perform and optimistic concurrency update using the `if-match` header.
+    /// `current` will fetch the current value for the key and store the etag
```
Not sure why this and other doc comments needed reformatting, or is it a new lint?

kate-goldenring replied:
I usually lint at 80 chars so must have updated this by accident. looks like we use 100.

itowlson reviewed (commented):
I think my comments mostly reflect my ignorance rather than actual issues.  Hopefully @devigned can provide a more informed review!

rylev commented on crates/key-value-azure/Cargo.toml:
```diff
@@ -10,9 +10,9 @@ rust-version.workspace = true
-azure_data_cosmos = { git = "https://github.com/azure/azure-sdk-for-rust.git", rev = "8c4caa251c3903d5eae848b41bb1d02a4d65231c" }
-azure_identity = { git = "https://github.com/azure/azure-sdk-for-rust.git", rev = "8c4caa251c3903d5eae848b41bb1d02a4d65231c" }
-azure_core = { git = "https://github.com/azure/azure-sdk-for-rust.git", rev = "8c4caa251c3903d5eae848b41bb1d02a4d65231c" }
+azure_data_cosmos = { git = "https://github.com/azure/azure-sdk-for-rust.git", tag = "azure_data_cosmos-0.21.0"  }
```
Can we use the crates.io release instead?

devigned replied:
Looks like 0.22.0 is available on crates.

kate-goldenring replied:
I tried to update these changes to be on top of 0.22.0 when it came out a couple days ago but they removed some functionality for performing swaps using the `if-match` header. I filed an issue here: https://github.com/Azure/azure-sdk-for-rust/issues/2254

I can switch back to pointing to a commit sha if that is more robust

lann replied:
Is there some specific reason we need to use the repo here?

```suggestion
azure_data_cosmos = "0.21"
```

kate-goldenring replied:
Sorry for the confusion here. I am not sure why i thought it wasn't published to crates. Updated

rylev commented on crates/key-value-azure/src/store.rs:
```diff
@@ -158,24 +159,35 @@ impl Store for AzureCosmosStore {
         self.client
             .create_document(pair)
             .is_upsert(true)
+            .allow_tentative_writes(TentativeWritesAllowance::Allow)
```
We should add a comment why this is needed.

rylev commented on crates/key-value-azure/src/store.rs:
```diff
@@ -216,24 +228,67 @@ impl Store for AzureCosmosStore {
         Ok(())
     }
 
+    // WARNING: this function only works on the primary region because the
```
Nit: I think it's preferable to make this a doc comment (rather than a plain comment) so it's more easily visible outside of the source code.

rylev commented on crates/key-value-azure/src/store.rs:
```diff
@@ -216,24 +228,67 @@ impl Store for AzureCosmosStore {
-                    .to_string(),
-            )),
+        {
+            Err(e) => {
+                if e.as_http_error()
+                    .map(|e| e.status())
```
It might help with @itowlson's comment directly below this to do something like:

```suggestion
            Err(e)  if e.as_http_error().map(|e| e.status() == 404).umwrap_or(false) => {
               
```

rylev reviewed (commented):
Looking good! I think it's by and large ready to merge, but it would be nice to see the nits addressed.

devigned commented on crates/key-value-azure/src/store.rs:
```diff
@@ -327,15 +383,18 @@ impl Cas for CompareAndSwap {
                 // attempt to replace the document if the etag matches
                 doc_client
                     .replace_document(pair)
+                    .allow_tentative_writes(TentativeWritesAllowance::Allow)
```
I'm a little concerned that by allowing tentative writes, this becomes non-atomic via last in wins default conflict resolution. It seems like this operation should be targeted at the hub rather than one of the spoke db instances. Thoughts?

kate-goldenring replied:
This is a good point. One option would be to remove `allow_tentative_writes` from all atomic operations. Right now, through runtime config, there is no way to configure a target endpoint/region so the hub endpoint is always generated anyways. This brings up the point that we should probably hold off on `allow_tentative_writes` until we update runtime config to support configuring non-primary regions.

devigned reviewed (commented):
I really like the improvements. I did mention concern about tentative writes in atomic operations. In the scenario where a swap occurs at roughly the same time on the same key, you could imagine 2 processes running and fetching the same etag version of the document, then the first committing the updated document to the hub while the second is tentatively committing the document to a spoke instance after the first. The second document committed seems like it would win with the [default conflict resolution protocol of last in wins](https://learn.microsoft.com/en-us/azure/cosmos-db/conflict-resolution-policies). 

I'm not sure if this is the right answer, but we may want to consider sending atomic reads and writes to the hub db instance and avoid multi-region writes.

kate-goldenring commented:
Updated the PR to only contain the set, delete, and get fixes. Removed non-primary region support since there is no way to configure it in runtime config

itowlson commented on crates/key-value-azure/src/store.rs:
```diff
@@ -138,14 +138,18 @@ struct AzureCosmosStore {
     /// An optional store id to use as a partition key for all operations.
     ///
-    /// If the store id not set, the store will use `/id` as the partition key.
+    /// If the store ID is not set, the store will use `/id` (the row key) as
```
Thank you Kate, this is so much clearer!

itowlson commented:
@kate-goldenring Do you have a moment to resolve the merge conflict on this?  Sorry I think there has been a bit of churn while GH actions was malingering

kate-goldenring commented:
Thanks for the bump @itowlson

itowlson commented:
Bypassing checks as only flakes and bads failing now
